[CALCITE-5935] Add CODE_POINTS_TO_BYTES function (enabled in BigQuery…

macroguo-ghy opened the pull request:
… library)

julianhyde commented on babel/src/test/resources/sql/big-query.iq:
```diff
@@ -3593,5 +3593,61 @@ FROM items;
+SELECT CODE_POINTS_TO_BYTES(array[-1]) as result;
+Input arguments of CODE_POINTS_TO_BYTES out of range: -1
+!error
+
```
+1

julianhyde commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -816,6 +816,33 @@ public static String charFromUtf8(int n) {
+   * SQL CODE_POINTS_TO_BYTES function.
+   */
+  public static @Nullable ByteString codePointsToBytes(@Nullable List codePoints) {
+    if (codePoints == null) {
```
I think you can and should remove the `@Nullable` from `codePoints`. The function should not be called if `codePoints` is null.

julianhyde commented on site/_docs/reference.md:
```diff
@@ -2681,6 +2681,7 @@ BigQuery's type system uses confusingly different names for types and functions:
 | b | CEIL(value)                                    | Similar to standard `CEIL(value)` except if *value* is an integer type, the return type is a double
 | m s | CHAR(integer)                                | Returns the character whose ASCII code is *integer* % 256, or null if *integer* &lt; 0
 | b o p | CHR(integer)                               | Returns the character whose UTF-8 code is *integer*
+| b | CODE_POINTS_TO_BYTES(array&lt;integer&gt;)     | Converts an *array&lt;integer&gt;* of extended ASCII code points to bytes, return null if array is null or the element of the array is null
```
I would call the argument `integers`, and describe it 'Converts *integers*, an array of integers between 0 and 255, into a byte string'. `array<integer>` is not how Calcite writes types.

I would mention that it throws if any element is not between 0 and 255 inclusive; no need to mention that it returns null if `integers` is null.

julianhyde reviewed (changes requested):
A few nits, but looks good.

macroguo-ghy commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -816,6 +816,33 @@ public static String charFromUtf8(int n) {
+    int length = codePoints.size();
+    byte[] bytes = new byte[length];
+    for (int i = 0; i < length; i++) {
+      Object codePoint = codePoints.get(i);
```
because of `bytes[i] = (byte) cp;`, I need to know the index in the loop for byte array assignment.

macroguo-ghy commented:
The CI in Windows (JDK 17) has failed, but I don't think it fails because of this commit.

herunkang2018 commented on core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:
```diff
@@ -510,6 +510,30 @@ public static SqlOperandTypeChecker variadic(
+          return opName + "(<INTEGER ARRAY>)";
+        }
+      };
+
```
Would you like to remove useless blank line?

macroguo-ghy replied:
Done.

herunkang2018 commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -831,6 +831,30 @@ public static String charFromUtf8(int n) {
+      if (codePoint instanceof Number) {
+        long cp = ((Number) codePoint).longValue();
+        if (cp < 0 || cp > 255) {
+          throw new IllegalArgumentException(
```
I would use `CalciteResource` to store the exception string, this is a more uniform approach.

macroguo-ghy replied:
Done.

herunkang2018 commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -1804,6 +1804,28 @@ void testCastToBoolean(CastType castType, SqlOperatorFixture f) {
+    f.checkFails("code_points_to_bytes(array[2147483648, 1])",
+        "Input arguments of CODE_POINTS_TO_BYTES out of range: 2147483648", true);
+
+    f.checkString("code_points_to_bytes(array[65,66,67,68])", "41424344", "VARBINARY NOT NULL");
```
Minor comment for code format, I would use `array[65, 66, 67, 68]` instead `array[65,66,67,68]` for better code readability.

macroguo-ghy replied:
Done.

herunkang2018 commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -1804,6 +1804,28 @@ void testCastToBoolean(CastType castType, SqlOperatorFixture f) {
+    final SqlOperatorFixture f = fixture()
+        .setFor(SqlLibraryOperators.CODE_POINTS_TO_BYTES, VM_FENNEL, VM_JAVA)
+        .withLibrary(SqlLibrary.BIG_QUERY);
+    f.checkFails("^code_points_to_bytes('abc')^",
```
Would you like to change the `'abc'` to an array literal?

macroguo-ghy replied:
Good suggestion! I add a new test case for string literals array input.

herunkang2018 commented on site/_docs/reference.md:
```diff
@@ -2681,6 +2681,7 @@ BigQuery's type system uses confusingly different names for types and functions:
 | b | CEIL(value)                                    | Similar to standard `CEIL(value)` except if *value* is an integer type, the return type is a double
 | m s | CHAR(integer)                                | Returns the character whose ASCII code is *integer* % 256, or null if *integer* &lt; 0
 | b o p | CHR(integer)                               | Returns the character whose UTF-8 code is *integer*
+| b | CODE_POINTS_TO_BYTES(integers)                 | Converts *integers*, an array of integers between 0 and 255 inclusive, throws error if any element is out of range
```
Maybe a small typo, `an array of integers` -> `an array of integer`

macroguo-ghy replied:
I think `an array of integers`· is the correct syntax.

macroguo-ghy commented:
I ran a Github Action for this commits in my repository (https://github.com/macroguo-ghy/calcite/pull/2) and it passed all CI checks, including JDK17 in Windows. I believe the weird failures may be caused by gradle cache. Should I open a new PR to address this problem?

macroguo-ghy commented:
Hi @julianhyde @tanclary @herunkang2018 , thanks for your review and suggestions! I cannot solve weird failures for CI / Windows (JDK 17) in this PR, so I have opened a new PR (https://github.com/apache/calcite/pull/3389) that has passed all checks. This new PR contains the same changes, but I have rebased and squashed the commits.
I will close this PR later, so if you have any further suggestions,  please feel free to comment on the new PR:)
